include/ensmallen_bits/wn_grad/wn_grad.hpp

```diff
@@ -0,0 +1,133 @@
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_WN_GRAD_WN_GRAD_HPP
+#define ENSMALLEN_WN_GRAD_WN_GRAD_HPP
+
+#include <ensmallen_bits/sgd/sgd.hpp>
+#include "wn_grad_update.hpp"
+
+namespace ens {
+
+/**
+ * WNGrad is a general nonlinear update rule for the learning rate, so that
+ * the learning rate can be initialized at a high value, and is subsequently
+ * decreased according to gradient observations.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{Wu2018,
+ *   author  = {{Wu}, X. and {Ward}, R. and {Bottou}, L.},
+ *   title   = {WNGrad: Learn the Learning Rate in Gradient Descent},
+ *   journal = {ArXiv e-prints},
+ *   year    = {2018},
+ *   url     = {https://arxiv.org/abs/1803.02865},
+ * }
+ * @endcode
+ *
+ * For WNGrad to work, a DecomposableFunctionType template parameter is
+ * required. This class must implement the following function:
+ *
+ *   size_t NumFunctions();
+ *   double Evaluate(const arma::mat& coordinates,
+ *                   const size_t i,
+ *                   const size_t batchSize);
+ *   void Gradient(const arma::mat& coordinates,
+ *                 const size_t i,
+ *                 arma::mat& gradient,
+ *                 const size_t batchSize);
+ *
+ * NumFunctions() should return the number of functions (\f$n\f$), and in the
+ * other two functions, the parameter i refers to which individual function (or
+ * gradient) is being evaluated.  So, for the case of a data-dependent function,
+ * such as NCA, NumFunctions() should return the number of points in the
+ * dataset, and Evaluate(coordinates, 0) will evaluate the objective function on
+ * the first point in the dataset (presumably, the dataset is held internally in
+ * the DecomposableFunctionType).
+ */
+class WNGrad
+{
+ public:
+  /**
+   * Construct the WNGrad optimizer with the given function and parameters. The
+   * defaults here are not necessarily good for the given problem, so it is
+   * suggested that the values used be tailored to the task at hand.  The
+   * maximum number of iterations refers to the maximum number of points that
+   * are processed (i.e., one iteration equals one point; one iteration does not
+   * equal one pass over the dataset).
+   *
+   * @param stepSize Step size for each iteration.
+   * @param batchSize Number of points to process in a single step.
+   * @param maxIterations Maximum number of iterations allowed (0 means no
+   *        limit).
+   * @param tolerance Maximum absolute tolerance to terminate algorithm.
+   * @param shuffle If true, the function order is shuffled; otherwise, each
+   *        function is visited in linear order.
+   */
+  WNGrad(const double stepSize = 0.56,
```

Comment:
Should we go with the paper's value of 0.562?  (I know it does not make much difference empirically, but someone looking through the code (like me) might get a bit confused otherwise.)

Reply:
Good point.